packages/ember-data/lib/system/relationships/state/has_many.js

```diff
@@ -139,12 +139,13 @@ ManyRelationship.prototype.computeChanges = function(records) {
   for (i = 0; i < length; i++) {
     record = records[i];
     //Need to preserve the order of incoming records
-    if (hasManyArray.objectAt(i) === record ) {
+    if (hasManyArray.objectAt(i) === record) {
       continue;
     }
     this.removeCanonicalRecord(record);
     this.addCanonicalRecord(record, i);
   }
+  this.flushCanonicalLater();
```

Comment:
`this.addCanonicalRecord` will also call `flushCanonicalLater `. This case is makes sure we don't run `flushCanonical` more then once it a given run loop. `this.flushCanonicalLater();` was added here because its possible for `addCanonicalRecord` to not be because a record was in the correct position but its state has been updated so the `flushCanonical` still needs to be re-run for the relationship to render correctly.